File: 20.valid-parentheses/solution.cpp
class Solution {
public:
	bool isValid(string s) {
		stack<char> st;
		for (string::iterator it = s.begin(); it != s.end(); it++)
		{
			char c = *it;
			if (c == '{' || c == '[' || c == '(')
				st.push(*it);
			else
			{
				if (st.empty())
					return false;
				char vl = c - st.top();
				if (vl != 1 && vl != 2)
					return false;
				st.pop();
			}
		}
		return st.empty();
	}
};

File: 20.valid-parentheses/solution.kt
class Solution {
    fun isValid(s: String): Boolean {
        var stack = ArrayList<Char>()
        for (ch in s) {
            when (ch) {
                '(' -> stack.add('(')
                ')' -> if (stack.isEmpty() || '(' != stack.removeAt(stack.size - 1)) return false
                '[' -> stack.add('[')
                ']' -> if (stack.isEmpty() || '[' != stack.removeAt(stack.size - 1)) return false
                '{' -> stack.add('{')
                '}' -> if (stack.isEmpty() || '{' != stack.removeAt(stack.size - 1)) return false
            }
        }
        return stack.isEmpty()
    }
}
